package/MDAnalysis/analysis/diffusionmap.py

```diff
@@ -0,0 +1,432 @@
+    """Non-linear dimension reduction method
+
+    Dimension reduction with diffusion mapping of selected structures in a
+    trajectory.
+
+    Attributes
+    ----------
+    eigenvalues: array
+        Eigenvalues of the diffusion map
+    eigenvectors: array
+        Eigenvectors of the diffusion map
+    diffusion_space : array
+        After calling `transform(num_eigenvectors)` the diffusion map embedding
+        into the lower dimensional diffusion space will exist here.
+
+    Methods
+    -------
+    run()
+        Constructs an anisotropic diffusion kernel and performs eigenvalue
+        decomposition on it.
+    transform(num_eigenvectors)
+        Perform an embedding of a frame into the eigenvectors representing
+        the collective coordinates.
+    """
+
+    def __init__(self, u, epsilon=1, manifold_density=None,
+                 timescale=1, force=False, **kwargs):
+        """
+        Parameters
+        -------------
+        u : MDAnalysis Universe or DistanceMatrix object
+            Can be a Universe, in which case one must supply kwargs for the
+            initialization of a DistanceMatrix. Otherwise, this can be a
+            DistanceMatrix already initialized. Either way, this will be made
+            into a diffusion kernel.
+        epsilon : Float
+            Specifies the method used for the choice of scale parameter in the
+            diffusion map. More information in [1], [2] and [3], Default: 1.
+        manifold_density: list, optional
+            The list has to have the same length as the trajectory.
+            With 'None' the weight of each frame of the trajectory will be the
+            same, Default : None
+        timescale: int, optional
+            The number of steps in the random walk, large t reflects global
+            structure whereas small t indicates local structure, Default: 1
+        force : boolean, optional
+            A boolean to override an exception thrown for a large
+            DistanceMatrix, Default: False
+        """
+        if isinstance(u, mda.Universe):
+            select = kwargs.get('select', "all")
+            metric = kwargs.get('metric', rmsd)
+            cutoff = kwargs.get('cutoff', 1E0-5)
+            weights = kwargs.get('weights')
+            start = kwargs.get('start')
+            stop = kwargs.get('stop')
+            step = kwargs.get('step')
+            self._dist_matrix = DistanceMatrix(u,select, metric, cutoff,
+                                               weights, start, stop, step)
+        elif isinstance(u, DistanceMatrix):
+            self._dist_matrix = u
+        else:
+            raise ValueError("U is not a Universe or DistanceMatrix and"
+                             " so the DiffusionMap has no data to work with.")
+        self._epsilon = epsilon
+        # important for transform function and length of .run() method
+        self._nframes = self._dist_matrix.nframes
+        if self._nframes > 2000 and not force:
```

Comment:
Oh maybe not! I managed to crash my laptop the other day doing an
eigenvalue decomposition. I'll try to break things tomorrow.
On Wed, Jun 29, 2016 at 10:54 PM David Dotson <notifications@github.com>
wrote:

> In package/MDAnalysis/analysis/diffusionmap.py
> <https://github.com/MDAnalysis/mdanalysis/pull/863#discussion_r69075107>:
>
> > +            cutoff = kwargs.get('cutoff', 1E0-5)
> > +            weights = kwargs.get('weights')
> > +            start = kwargs.get('start')
> > +            stop = kwargs.get('stop')
> > +            step = kwargs.get('step')
> > +            self._dist_matrix = DistanceMatrix(u,select, metric, cutoff,
> > +                                               weights, start, stop, step)
> > +        elif isinstance(u, DistanceMatrix):
> > +            self._dist_matrix = u
> > +        else:
> > +            raise ValueError("U is not a Universe or DistanceMatrix and"
> > +                             " so the DiffusionMap has no data to work with.")
> > +        self._epsilon = epsilon
> > +        # important for transform function and length of .run() method
> > +        self._nframes = self._dist_matrix.nframes
> > +        if self._nframes > 2000 and not force:
>
> Is this really necessary? Python will raise a MemoryError on allocation
> of a large array (which will happen in DistanceMatrix._prepare) if it is
> too large to fit in memory.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/MDAnalysis/mdanalysis/pull/863/files/a76d44b5ecc177778285e60218a1d71715d333fc#r69075107>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AKcARiLGnzFnqxxIstjd3bjeA7Q7qdKsks5qQ1oJgaJpZM4IpYIS>
> .
>
-- 
Have a wonderful day,
John J. Detlefs